File: tools/boot_check.py
#!/usr/bin/env python3
"""Cold-boot the built ROM in Mesen and save a screenshot.

Mesen's test runner needs absolute paths, and its Lua script writes the image
itself, so this wrapper just resolves paths and reports where the shot landed.

  tools/boot_check.py [--frames 900] [--out build/reports/boot.png]
"""

from __future__ import annotations

import argparse
import os
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
MESEN = Path("/Applications/Mesen.app/Contents/MacOS/Mesen")
DEFAULT_ROM = ROOT / "build" / "srw4-th-test.sfc"
LUA = ROOT / "tools" / "lua" / "boot-check.lua"


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--frames", type=int, default=900)
    parser.add_argument("--rom", type=Path, default=DEFAULT_ROM)
    parser.add_argument("--out", type=Path, default=ROOT / "build" / "reports" / "boot.png")
    parser.add_argument("--timeout", type=int, default=120)
    args = parser.parse_args()

    if not MESEN.exists():
        print(f"Mesen not found at {MESEN}", file=sys.stderr)
        return 1
    rom = args.rom.resolve()
    if not rom.exists():
        print(f"no build to check: run tools/build.py first ({rom})", file=sys.stderr)
        return 1

    out = args.out.resolve()
    out.parent.mkdir(parents=True, exist_ok=True)
    if out.exists():
        out.unlink()

    env = dict(os.environ, SRW4_SHOT=str(out), SRW4_FRAMES=str(args.frames))
    subprocess.run(
        [
            str(MESEN),
            "--testRunner",
            f"--testRunnerTimeout={args.timeout}",
            "--noAudio",
            str(rom),
            str(LUA.resolve()),
        ],
        env=env,
        check=True,
    )

    if not out.exists():
        print("Mesen exited without writing a screenshot", file=sys.stderr)
        return 1
    print(f"{out.relative_to(ROOT)}  {out.stat().st_size} bytes  (frame {args.frames})")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/lua/boot-check.lua
-- P0 boot check: run the built ROM from a cold boot, press start a few times,
-- and save a screenshot. Proves the expanded image still runs; it is not a
-- substitute for judging text, which must always come from a live redraw.
local OUT = os.getenv("SRW4_SHOT") or "build/reports/boot.png"
local LAST = tonumber(os.getenv("SRW4_FRAMES") or "900")

local frame = 0

local function onInput()
  frame = frame + 1
  if frame > 60 and frame % 30 == 0 then
    emu.setInput({ start = true }, 0)
  end
end

local function onFrame()
  if frame >= LAST then
    local png = emu.takeScreenshot()
    local f = io.open(OUT, "wb")
    f:write(png)
    f:close()
    emu.log("screenshot written to " .. OUT .. " at frame " .. frame)
    emu.stop(0)
  end
end

emu.addEventCallback(onInput, emu.eventType.inputPolled)
emu.addEventCallback(onFrame, emu.eventType.endFrame)
